### ISSUE TITLE
Partially typed row iterator?

### ISSUE THREAD

**Issue by tkf** [JuliaData/Tables.jl#137]
Would it make sense to introduce a utility function in Tables.jl for creating "partially typed" row iterator?  The idea is to iterate over element like this

```julia
struct PartiallyTypedRow{names,types,Typed<:NamedTuple{names,types},Rest}
    typed::Typed
    rest::Rest
end

function Base.getproperty(row::PartiallyTypedRow, name::Symbol)
    haskey(getfield(row, :typed), name) && return getfield(row, :typed)[name]
    return getproperty(getfield(row, :rest), name)
end
```

The expectation is that this would be beneficial if the iteration involves some kind of filtering. Then, we can balance execution and compilation time by only specializing for the columns that are required for processing up to the filtering.

Here is a toy example:

```julia
using BenchmarkTools
using DataFrames
using PartiallyTypedTables: partiallytyped
using Tables

# Monkey patch!
Tables.rows(df::DataFrames.AbstractDataFrame) = eachrow(df)

function filtered_sum(rows)
    s = 0.0
    for r in rows
        if r.a > 3
            s += r.b
        end
    end
    return s
end

SUITE = BenchmarkGroup()

n = 1_000_000
df = DataFrame(a = randn(n), b = randn(n))

SUITE["none"] = @benchmarkable filtered_sum($(eachrow(df)))
SUITE["partial"] = @benchmarkable filtered_sum($(partiallytyped(df, :a)))
SUITE["full"] = @benchmarkable filtered_sum($(Tables.rows(DataFrames.columntable(df))))
```

Running this benchmark shows:

```
  "partial" => BenchmarkTools.Trial:
          memory estimate:  30.60 MiB
          allocs estimate:  1005444
          --------------
          minimum time:     9.503 ms (0.00% GC)
          median time:      14.722 ms (34.26% GC)
          mean time:        13.523 ms (27.03% GC)
          maximum time:     25.528 ms (0.00% GC)
          --------------
          samples:          367
          evals/sample:     1
  "none" => BenchmarkTools.Trial:
          memory estimate:  76.37 MiB
          allocs estimate:  4004933
          --------------
          minimum time:     79.398 ms (6.00% GC)
          median time:      85.665 ms (11.42% GC)
          mean time:        85.286 ms (10.04% GC)
          maximum time:     110.665 ms (4.87% GC)
          --------------
          samples:          59
          evals/sample:     1
  "full" => BenchmarkTools.Trial:
          memory estimate:  0 bytes
          allocs estimate:  0
          --------------
          minimum time:     963.316 μs (0.00% GC)
          median time:      1.014 ms (0.00% GC)
          mean time:        1.036 ms (0.00% GC)
          maximum time:     1.967 ms (0.00% GC)
          --------------
          samples:          4815
          evals/sample:     1
```

Indeed, partially typed iterator is much faster than bare `eachrow` iterator but not as fast as the fully specialized iterator.

Here is the [full implementation](https://github.com/tkf/PartiallyTypedTables.jl/blob/8844d0dad0f0fb14480563ad9acb774fcc9755a4/src/PartiallyTypedTables.jl) and the [benchmark](https://github.com/tkf/PartiallyTypedTables.jl/blob/8844d0dad0f0fb14480563ad9acb774fcc9755a4/benchmark/benchmarks.jl).

cc @bkamins

**Comment by bkamins** [comment]
What I typically do in such cases now is:
```
Tables.rows(DataFrames.columntable(df[!, [:a, :b]))
```
However, what you propose is a nice idea. My thought is that it might be even better to improve `eachrow` and `DataFrameRow` in DataFrames.jl than adding it here. Then one could call something like `eachrow(df, typed=[:a, :b])`. What do you think?

CC @nalimilan

**Comment by tkf** [comment]
Right, my benchmark is not really a good motivating example.  What I was thinking was for the case you need some operation on "wide row" after the filtering.  Something like

```julia
idxs :: Vector{Symbol}  # a long vector
s = zeros(length(idxs))
for r in rows
    if rare_condition(r.a)
        s .+= r[idxs] ./ mean(r[idxs])
    end
end
```

> My thought is that it might be even better to improve `eachrow` and `DataFrameRow` in DataFrames.jl than adding it here.

Given that DataFrames.jl is the canonical example of "untyped" table, I guess this is practically fine. But, IIUC, Tables.jl 1.0 also is going to have Dict-based table which can also be as untyped as DataFrame. Because of this, Tables.jl may be a good place to have this.

(It was also nice to see that I could implement it in terms of Tables.jl)

> Then one could call something like `eachrow(df, typed=[:a, :b])`.

I think having a uniform entry point like this is great for DataFrames.jl users.  But I guess this can just directly call `Tables.row`, right?  (Though I guess you need to refactor the code as, IIUC, `Tables.rows` calls `eachrow` now.)

**Comment by quinnj** [comment]
We essentially have a version of this with `Tables.eachcolumn` methods right now; they are generated functions that switch strategy based on how wide the schema is, even doing a run-length encoding on types as a strategy. I've also had thoughts on having DataFrames do something like that, since we keep having this tension between wanting the performance of strong-typed in some places, yet wanting to be a good "production" player in case extremely wide tables are encountered.

With that strategy, we would investigate certain thresholds where we would do `columntable(df)` (smallest cases), run-length encoding, or untyped `DataFrameRow` (widest cases). I definitely think this is a cool/interesting area to play around with. I've always thought it would be something to play around with in DataFrames, but it might be interesting if we can make something general.

**Comment by tkf** [comment]
> We essentially have a version of this with `Tables.eachcolumn` methods right now;

IIUC, it seems that they are specialized for schema. But since type parameters can be large for wide tables, I suppose its compilation overhead scales with the width of the table? What I'm suggesting here is to bring only a subset of "schema" to the type-domain. So, isn't it a different approach?

> With that strategy, we would investigate certain thresholds where we would do `columntable(df)` (smallest cases), run-length encoding, or untyped `DataFrameRow` (widest cases).

Are you thinking an "automagic" solution based on some kind of heuristics?  What I have in mind is more like a basic building block for manually controlling a subset of columns to be typed.  I think this could be useful when you know what columns are accessed by "construction" of the computation.  For example, macro-based frontends like DataFramesMeta.jl and Query.jl know such information.  I think it's nice to have both automagic highl-level API and a transparent low-level layer.

BTW, I didn't know about run-length encoding in `eachcolumn`. That's extremely cool :)

**Comment by quinnj** [comment]
Ok, I understand what the original ask is here better now; yeah, I think this sounds cool and could be a useful utility for avoiding over-compilation.
